src/gui/touchscreengui.cpp

```diff
@@ -107,6 +107,9 @@ static EKEY_CODE id_to_keycode(touch_gui_button_id id)
 			break;
 	}
 	assert(!key.empty());
+	if (g_settings->get("keymap_" + key) == "")
+		return KEY_NONE;
+
 	return keyname_to_keycode(g_settings->get("keymap_" + key).c_str());
```

Comment:
This still throws an `UnknownKeycode` exception if the key name is invalid.

A possible solution:

```patch
diff --git a/src/client/keycode.cpp b/src/client/keycode.cpp
index 1fcd15e01b..14360b1a1c 100644
--- a/src/client/keycode.cpp
+++ b/src/client/keycode.cpp
@@ -18,7 +18,6 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 */
 
 #include "keycode.h"
-#include "exceptions.h"
 #include "settings.h"
 #include "log.h"
 #include "debug.h"
@@ -26,13 +25,6 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/string.h"
 #include "util/basic_macros.h"
 
-class UnknownKeycode : public BaseException
-{
-public:
-	UnknownKeycode(const char *s) :
-		BaseException(s) {};
-};
-
 struct table_key {
 	const char *Name;
 	irr::EKEY_CODE Key;
diff --git a/src/client/keycode.h b/src/client/keycode.h
index 7036705d16..6bc44bb1d4 100644
--- a/src/client/keycode.h
+++ b/src/client/keycode.h
@@ -19,11 +19,19 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
 #pragma once
 
+#include "exceptions.h"
 #include "irrlichttypes.h"
 #include "Keycodes.h"
 #include <IEventReceiver.h>
 #include <string>
 
+class UnknownKeycode : public BaseException
+{
+public:
+	UnknownKeycode(const char *s) :
+		BaseException(s) {};
+};
+
 /* A key press, consisting of either an Irrlicht keycode
    or an actual char */
 
diff --git a/src/gui/touchscreengui.cpp b/src/gui/touchscreengui.cpp
index eb73008879..b7f034cda2 100644
--- a/src/gui/touchscreengui.cpp
+++ b/src/gui/touchscreengui.cpp
@@ -107,10 +107,18 @@ static EKEY_CODE id_to_keycode(touch_gui_button_id id)
 			break;
 	}
 	assert(!key.empty());
-	if (g_settings->get("keymap_" + key) == "")
-		return KEY_NONE;
 
-	return keyname_to_keycode(g_settings->get("keymap_" + key).c_str());
+	std::string resolved = g_settings->get("keymap_" + key);
+	EKEY_CODE code;
+	try {
+		code = keyname_to_keycode(resolved.c_str());
+	} catch (UnknownKeycode &e) {
+		code = KEY_KEY_CODES_COUNT;
+		if (!resolved.empty())
+			warningstream << "TouchScreenGUI: Unknown key '" << resolved
+					<< "' for '" << key << "', ignoring." << std::endl;
+	}
+	return code;
 }
 
 static void load_button_texture(const button_info *btn, const std::string &path,
@@ -534,7 +542,7 @@ void TouchScreenGUI::init(ISimpleTextureSource *tsrc)
 	};
 	std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
 	for (const auto &pair : buttons) {
-		if (id_to_keycode(pair.first) == KEY_NONE)
+		if (id_to_keycode(pair.first) == KEY_KEY_CODES_COUNT)
 			continue;
 
 		const wchar_t* wide = converter.from_bytes(pair.second).c_str();
```

Reply:
Look good. Just moved `code` definition into beginning of the block

Reply:
Could we print warning even when var is empty? So people can have a clue why a TS button is missing?

Reply:
> Could we print warning even when var is empty? So people can have a clue why a TS button is missing?  
> edit: I would drop the `if (!resolved.empty())` condition if you wouldn't mind

Makes sense. You could make the message even more helpful by replacing "ignoring" with "hiding button".